keps/sig-scheduling/4671-gang-scheduling/README.md

```diff
@@ -0,0 +1,932 @@
+
+Workloads that require gang scheduling often also need all members of the gang to be as topologically "close" to one another as possible, in order to perform adequately. Existing Pod affinity rules influence pod placement, but they do not consider the gang as a unit of scheduling and they do not cause the scheduler to efficiently try multiple mutually exclusive placement options for a set of pods. The design of the Workload object introduced in this KEP anticipates how Gang Scheduling support can evolve over subsequent KEPs into full Topology-aware scheduling support in kube-scheduler.
+
+The `Workload` object will allow kube-scheduler to be aware that pods are part of workloads with complex internal structure.  Those workloads include builtins like `Job` and `StatefulSet`, and custom workloads, like `JobSet`, `LeaderWorkerSet`, `MPIJob` and `TrainJob`. All of these workload types are used for AI training and inference use cases.
+
+
+### Goals
+- Introduce a concept of a `Workload` as a primary building block for workload-aware scheduling vision
+- Implement the first version of `Workload` API necessary for defining a Gang
+- Ensuring that we can extend `Workload` API in backward compatible way toward north-star API
+- Ensuring that `Workload` API will be usable for both built-in and third-party workload controllers and APIs
+- Implement first version of gang-scheduling in kube-scheduler
+- Provide full backward compatibility for all existing scheduling features
+
+### Non-Goals
+
+- Take away responsibility to create pods from controllers.
+- Bring fairness or multiple workload queues in kube-scheduler. Kueue and Volcano.sh will continue to provide this.
+- Map all the declarative state and behaviors into `Workload` object. It is focused only on scheduling-related parts.
+
+The following are non-goals for this KEP but will probably soon appear to be goals for follow-up KEPs:
+
+- Introduce a concept of `Reservation` that can be later consumed by pods.
+- Workload-level preemption.
+- Address resource contention between different schedulers (including possible deadlocks).
+- Address the problem of premature preemptions in case the higher priority workloads does not
+  eventually schedule.
+
+
+## Proposal
+
+The `spec.workload` field will be added to the Pod resource.  A sample pod with this new field looks like this:
+```yaml
+apiVersion: v1
+kind: Pod
+spec:
+  ...
+  workload:
+    name: job-1
+  ...
+```
+
+The above pod might be one of several pods created by a `Job` like this.
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: job-1
+spec:
+  completions: 100
+  parallelism: 100
+  template:
+    spec:
```

Comment:
If my understanding is correct we don't be touching the Job controller at this stage so the user would need to also provide the Workload as follow, no ?

```suggestion
    spec:
      workload:
        name: job-1
```